File: mobile/lib/core/services/vitals_service.dart
import '../models/vital_measurement_model.dart';
import '../mappers/vital_mapper.dart';
import 'api_service.dart';

class VitalsService {
  final ApiService _apiService = ApiService();

  void _log(String message) {
    print('🔍 [VITALS] $message');
  }

  // Get all vitals for a user
  Future<List<VitalMeasurementModel>> getVitals(
    String userId, {
    String? elderUserId,
  }) async {
    _log('📋 Fetching vitals for user: $userId');
    try {
      final response = await _apiService.get(
        '/vitals',
        queryParameters:
            elderUserId != null ? {'elderUserId': elderUserId} : null,
      );

      if (response.statusCode == 200) {
        final List<dynamic> data = response.data is List ? response.data : [];
        final vitals = data
            .map((json) => VitalMapper.fromApiResponse(
                json is Map<String, dynamic> ? json : Map<String, dynamic>.from(json)))
            .toList()
          ..sort((a, b) => b.timestamp.compareTo(a.timestamp));
        _log('✅ Fetched ${vitals.length} vital measurements');
        return vitals;
      } else {
        _log('❌ Failed to fetch vitals: ${response.statusMessage}');
        throw Exception('Failed to fetch vitals: ${response.statusMessage}');
      }
    } catch (e) {
      _log('❌ Error fetching vitals: $e');
      throw Exception(e.toString());
    }
  }

  // Get vitals by type
Future<List<VitalMeasurementModel>> getVitalsByType(
  String userId,
  VitalType type, {
  String? elderUserId,
}) async {
    _log('📋 Fetching vitals by type: $type for user: $userId');
    try {
      // Convert type enum to string for API
      String typeStr;
      switch (type) {
        case VitalType.bloodPressure:
          typeStr = 'bloodPressure';
          break;
        case VitalType.bloodSugar:
          typeStr = 'bloodSugar';
          break;
        case VitalType.heartRate:
          typeStr = 'heartRate';
          break;
        case VitalType.temperature:
          typeStr = 'temperature';
          break;
        case VitalType.oxygenSaturation:
          typeStr = 'oxygenSaturation';
          break;
        case VitalType.weight:
          typeStr = 'weight';
          break;
      }

      final queryParameters = {'type': typeStr};
      if (elderUserId != null) {
        queryParameters['elderUserId'] = elderUserId;
      }

      final response = await _apiService.get(
        '/vitals',
        queryParameters: queryParameters,
      );

      if (response.statusCode == 200) {
        final List<dynamic> data = response.data is List ? response.data : [];
        final vitals = data
            .map((json) => VitalMapper.fromApiResponse(
                json is Map<String, dynamic> ? json : Map<String, dynamic>.from(json)))
            .toList()
          ..sort((a, b) => b.timestamp.compareTo(a.timestamp));
        _log('✅ Fetched ${vitals.length} vital measurements of type $type');
        return vitals;
      } else {
        _log('❌ Failed to fetch vitals by type: ${response.statusMessage}');
        throw Exception('Failed to fetch vitals by type: ${response.statusMessage}');
      }
    } catch (e) {
      _log('❌ Error fetching vitals by type: $e');
      throw Exception(e.toString());
    }
  }

  // Add vital measurement
  Future<VitalMeasurementModel> addVital(VitalMeasurementModel vital) async {
    _log('➕ Adding vital measurement: ${vital.type}');
    try {
      final requestData = VitalMapper.toApiRequest(
        vital,
        elderUserId: vital.userId,
      );
      final response = await _apiService.post(
        '/vitals',
        data: requestData,
      );

      if (response.statusCode == 201 || response.statusCode == 200) {
        final addedVital = VitalMapper.fromApiResponse(response.data);
        _log('✅ Vital measurement added successfully');
        return addedVital;
      } else {
        _log('❌ Failed to add vital: ${response.statusMessage}');
        throw Exception('Failed to add vital: ${response.statusMessage}');
      }
    } catch (e) {
      _log('❌ Error adding vital: $e');
      throw Exception(e.toString());
    }
  }

  // Update vital
  Future<VitalMeasurementModel> updateVital(VitalMeasurementModel vital) async {
    _log('✏️ Updating vital measurement: ${vital.id}');
    try {
      final requestData = VitalMapper.toApiRequest(
        vital,
        elderUserId: vital.userId,
      );
      final response = await _apiService.patch(
        '/vitals/${vital.id}',
        data: requestData,
        queryParameters:
            vital.userId.isNotEmpty ? {'elderUserId': vital.userId} : null,
      );

      if (response.statusCode == 200) {
        final updatedVital = VitalMapper.fromApiResponse(response.data);
        _log('✅ Vital measurement updated successfully');
        return updatedVital;
      } else {
        _log('❌ Failed to update vital: ${response.statusMessage}');
        throw Exception('Failed to update vital: ${response.statusMessage}');
      }
    } catch (e) {
      _log('❌ Error updating vital: $e');
      throw Exception(e.toString());
    }
  }

  // Delete vital
  Future<void> deleteVital(String vitalId, {String? elderUserId}) async {
    _log('🗑️ Deleting vital measurement: $vitalId');
    try {
      final response = await _apiService.delete(
        '/vitals/$vitalId',
        queryParameters:
            elderUserId != null ? {'elderUserId': elderUserId} : null,
      );

      if (response.statusCode == 200) {
        _log('✅ Vital measurement deleted successfully');
      } else {
        _log('❌ Failed to delete vital: ${response.statusMessage}');
        throw Exception('Failed to delete vital: ${response.statusMessage}');
      }
    } catch (e) {
      _log('❌ Error deleting vital: $e');
      throw Exception(e.toString());
    }
  }

  // Get recent vitals (last 7 days)
  Future<List<VitalMeasurementModel>> getRecentVitals(
    String userId, {
    String? elderUserId,
  }) async {
    _log('📋 Fetching recent vitals (last 7 days) for user: $userId');
    try {
      final cutoffDate = DateTime.now().subtract(const Duration(days: 7));
      final response = await _apiService.get(
        '/vitals',
        queryParameters: {
          'startDate': cutoffDate.toIso8601String(),
          if (elderUserId != null) 'elderUserId': elderUserId,
        },
      );

      if (response.statusCode == 200) {
        final List<dynamic> data = response.data is List ? response.data : [];
        final vitals = data
            .map((json) => VitalMapper.fromApiResponse(
                json is Map<String, dynamic> ? json : Map<String, dynamic>.from(json)))
            .toList()
          ..sort((a, b) => b.timestamp.compareTo(a.timestamp));
        _log('✅ Fetched ${vitals.length} recent vital measurements');
        return vitals;
      } else {
        _log('❌ Failed to fetch recent vitals: ${response.statusMessage}');
        throw Exception('Failed to fetch recent vitals: ${response.statusMessage}');
      }
    } catch (e) {
      _log('❌ Error fetching recent vitals: $e');
      throw Exception(e.toString());
    }
  }

  // Calculate average for a vital type over a period
  Future<Map<String, dynamic>> calculateTrends(
    String userId,
    VitalType type, {
    int days = 7,
    String? elderUserId,
  }) async {
    _log('📊 Calculating trends for ${type.toString()} (last $days days)');
    try {
      String typeStr;
      switch (type) {
        case VitalType.bloodPressure:
          typeStr = 'bloodPressure';
          break;
        case VitalType.bloodSugar:
          typeStr = 'bloodSugar';
          break;
        case VitalType.heartRate:
          typeStr = 'heartRate';
          break;
        case VitalType.temperature:
          typeStr = 'temperature';
          break;
        case VitalType.oxygenSaturation:
          typeStr = 'oxygenSaturation';
          break;
        case VitalType.weight:
          typeStr = 'weight';
          break;
      }

      final queryParameters = {'kindCode': typeStr};
      if (elderUserId != null) {
        queryParameters['elderUserId'] = elderUserId;
      }

      final response = await _apiService.get(
        '/vitals/trends',
        queryParameters: queryParameters,
      );

      if (response.statusCode == 200) {
        final data = response.data;
        _log('✅ Trends calculated successfully');
        
        // Handle empty array response - calculate locally as fallback
        if (data is List && data.isEmpty) {
          _log('⚠️ Empty trends data from API, calculating locally as fallback');
          return await _calculateTrendsLocally(userId, type, days: days, elderUserId: elderUserId);
        }
        
        // Handle Map response
        if (data is Map<String, dynamic>) {
          final count = (data['count'] ?? 0) as int;
          // If API returns empty count, try local calculation as fallback
          if (count == 0) {
            _log('⚠️ API returned count=0, calculating locally as fallback');
            return await _calculateTrendsLocally(userId, type, days: days, elderUserId: elderUserId);
          }
          return {
            'average': (data['average'] ?? 0.0).toDouble(),
            'count': count,
            'hasAbnormal': (data['hasAbnormal'] ?? false) as bool,
            'measurements': (data['measurements'] ?? [])
                .map((json) => VitalMapper.fromApiResponse(
                      json is Map<String, dynamic>
                          ? json
                          : Map<String, dynamic>.from(json),
                    ))
                .toList(),
          };
        }
        
        // Fallback for unexpected data types - calculate locally
        _log('⚠️ Unexpected data type: ${data.runtimeType}, calculating locally as fallback');
        return await _calculateTrendsLocally(userId, type, days: days, elderUserId: elderUserId);
      } else {
        _log('❌ Failed to calculate trends: ${response.statusMessage}');
        throw Exception(
            'Failed to calculate trends: ${response.statusMessage}');
      }
    } 
    catch (e) {
      _log('❌ Error calculating trends from API: $e');
      _log('🔄 Falling back to local calculation');
      // Fallback to local calculation if API fails
      try {
        return await _calculateTrendsLocally(userId, type, days: days, elderUserId: elderUserId);
      } catch (localError) {
        _log('❌ Error in local calculation: $localError');
        throw Exception('Failed to calculate trends: ${e.toString()}');
      }
    }
  }

  // Calculate trends locally from vitals data
  Future<Map<String, dynamic>> _calculateTrendsLocally(
    String userId,
    VitalType type, {
    int days = 7,
    String? elderUserId,
  }) async {
    _log('📊 Calculating trends locally for ${type.toString()} (last $days days)');
    try {
      // Get all vitals for the user
      final allVitals = await getVitals(userId, elderUserId: elderUserId);
      
      // Filter by type and date range
      final cutoffDate = DateTime.now().subtract(Duration(days: days));
      final filteredVitals = allVitals.where((vital) {
        return vital.type == type && vital.timestamp.isAfter(cutoffDate);
      }).toList();

      if (filteredVitals.isEmpty) {
        _log('⚠️ No vitals found for local calculation');
        return {
          'average': 0.0,
          'count': 0,
          'hasAbnormal': false,
          'measurements': <VitalMeasurementModel>[],
        };
      }

      // Calculate average (handle blood pressure specially)
      double average = 0.0;
      if (type == VitalType.bloodPressure) {
        // For blood pressure, calculate average of systolic values
        final systolicValues = filteredVitals
            .map((v) {
              final parts = v.value.split('/');
              if (parts.length == 2) {
                return int.tryParse(parts[0]) ?? 0;
              }
              return 0;
            })
            .where((v) => v > 0)
            .toList();
        if (systolicValues.isNotEmpty) {
          average = systolicValues.reduce((a, b) => a + b) / systolicValues.length;
        }
      } else {
        // For other vitals, parse numeric value
        final numericValues = filteredVitals
            .map((v) => double.tryParse(v.value) ?? 0.0)
            .where((v) => v > 0)
            .toList();
        if (numericValues.isNotEmpty) {
          average = numericValues.reduce((a, b) => a + b) / numericValues.length;
        }
      }

      // Check for abnormal readings
      final hasAbnormal = filteredVitals.any((v) => v.isAbnormal());

      _log('✅ Local trends calculated: count=${filteredVitals.length}, average=$average');
      return {
        'average': average,
        'count': filteredVitals.length,
        'hasAbnormal': hasAbnormal,
        'measurements': filteredVitals,
      };
    } catch (e) {
      _log('❌ Error in local trends calculation: $e');
      return {
        'average': 0.0,
        'count': 0,
        'hasAbnormal': false,
        'measurements': <VitalMeasurementModel>[],
      };
    }
  }

  // Check for abnormal readings
  Future<List<VitalMeasurementModel>> getAbnormalReadings(
    String userId, {
    String? elderUserId,
  }) async {
    _log('⚠️ Fetching abnormal readings for user: $userId');
    try {
      final response = await _apiService.get(
        '/vitals/abnormal',
        queryParameters:
            elderUserId != null ? {'elderUserId': elderUserId} : null,
      );

      if (response.statusCode == 200) {
        final List<dynamic> data = response.data is List ? response.data : [];
        final vitals = data
            .map((json) => VitalMapper.fromApiResponse(
                json is Map<String, dynamic> ? json : Map<String, dynamic>.from(json)))
            .toList()
          ..sort((a, b) => b.timestamp.compareTo(a.timestamp));
        _log('✅ Fetched ${vitals.length} abnormal readings');
        return vitals;
      } else {
        _log('❌ Failed to fetch abnormal readings: ${response.statusMessage}');
        throw Exception('Failed to fetch abnormal readings: ${response.statusMessage}');
      }
    } catch (e) {
      _log('❌ Error fetching abnormal readings: $e');
      throw Exception(e.toString());
    }
  }

  // Get latest vitals per kind
  Future<Map<String, VitalMeasurementModel>> getLatestVitals(
    String userId, {
    String? elderUserId,
  }) async {
    _log('📋 Fetching latest vitals per kind for user: $userId');
    try {
      final response = await _apiService.get(
        '/vitals/latest',
        queryParameters:
            elderUserId != null ? {'elderUserId': elderUserId} : null,
      );

      if (response.statusCode == 200) {
        final data = response.data;
        final latestVitals = <String, VitalMeasurementModel>{};
        
        if (data is Map) {
          data.forEach((key, value) {
            if (value != null) {
              latestVitals[key] = VitalMapper.fromApiResponse(
                  value is Map<String, dynamic> ? value : Map<String, dynamic>.from(value));
            }
          });
        }
        
        _log('✅ Fetched latest vitals for ${latestVitals.length} types');
        return latestVitals;
      } else {
        _log('❌ Failed to fetch latest vitals: ${response.statusMessage}');
        throw Exception('Failed to fetch latest vitals: ${response.statusMessage}');
      }
    } catch (e) {
      _log('❌ Error fetching latest vitals: $e');
      throw Exception(e.toString());
    }
  }
}

File: backend/src/vitals/vitals.service.ts
import { Injectable, NotFoundException } from '@nestjs/common';
import { PrismaService } from '../prisma/prisma.service';
import { CreateVitalDto, VitalType } from './dto/create-vital.dto';
import { UpdateVitalDto } from './dto/update-vital.dto';
import { ActorContext } from '../common/services/access-control.service';

@Injectable()
export class VitalsService {
  constructor(private prisma: PrismaService) {}

  /**
   * Map app enum to database kindCode
   */
  private typeToKindCode(type: VitalType): string {
    const mapping: Record<VitalType, string> = {
      [VitalType.BLOOD_PRESSURE]: 'bp',
      [VitalType.BLOOD_SUGAR]: 'bs',
      [VitalType.HEART_RATE]: 'hr',
      [VitalType.TEMPERATURE]: 'temp',
      [VitalType.OXYGEN_SATURATION]: 'o2',
      [VitalType.WEIGHT]: 'weight',
    };
    return mapping[type] || type.toLowerCase();
  }

  /**
   * Map database kindCode to app enum
   */
  private kindCodeToType(kindCode: string): VitalType {
    const mapping: Record<string, VitalType> = {
      bp: VitalType.BLOOD_PRESSURE,
      bs: VitalType.BLOOD_SUGAR,
      hr: VitalType.HEART_RATE,
      temp: VitalType.TEMPERATURE,
      o2: VitalType.OXYGEN_SATURATION,
      weight: VitalType.WEIGHT,
    };
    return mapping[kindCode.toLowerCase()] || (kindCode as VitalType);
  }

  /**
   * Parse value string to value1/value2/valueText
   */
  private parseValue(type: VitalType, value: string): {
    value1: number | null;
    value2: number | null;
    valueText: string | null;
  } {
    if (type === VitalType.BLOOD_PRESSURE) {
      const parts = value.split('/');
      if (parts.length === 2) {
        return {
          value1: parseFloat(parts[0]) || null,
          value2: parseFloat(parts[1]) || null,
          valueText: null,
        };
      }
    }

    // For other types, try to parse as number
    const numValue = parseFloat(value);
    if (!isNaN(numValue)) {
      return {
        value1: numValue,
        value2: null,
        valueText: null,
      };
    }

    // If not a number, store as text
    return {
      value1: null,
      value2: null,
      valueText: value,
    };
  }

  /**
   * Format value1/value2/valueText to string
   */
  private formatValue(measurement: any): string {
    if (measurement.valueText) {
      return measurement.valueText;
    }

    if (measurement.value1 !== null && measurement.value2 !== null) {
      return `${measurement.value1}/${measurement.value2}`;
    }

    if (measurement.value1 !== null) {
      return measurement.value1.toString();
    }

    return '';
  }

  /**
   * Create vital measurement
   */
  async create(context: ActorContext, createDto: CreateVitalDto) {
    try {
      const { type, value, timestamp, notes } = createDto;
      
      // Ensure value is a string for parsing
      const valueStr = typeof value === 'string' ? value : String(value);
      const { value1, value2, valueText } = this.parseValue(type, valueStr);

      const measurement = await this.prisma.vitalMeasurement.create({
        data: {
          elderUserId: context.elderUserId,
          kindCode: this.typeToKindCode(type),
          unitCode: this.getUnitCode(type),
          value1: value1,
          value2: value2,
          valueText: valueText,
          recordedAt: new Date(timestamp),
          source: 'manual',
          notes: notes || null,
          recordedByUserId: context.actorUserId,
        },
      });

      return this.mapToResponse(measurement);
    } catch (error) {
      console.error('Error in VitalsService.create:', error);
      console.error('Context:', context);
      console.error('DTO:', createDto);
      throw error;
    }
  }

  /**
   * Find all vitals for a user
   */
  async findAll(
    context: ActorContext,
    type?: VitalType,
    startDate?: Date,
    endDate?: Date,
  ) {
    const where: any = {
      elderUserId: context.elderUserId,
    };

    if (type) {
      where.kindCode = this.typeToKindCode(type);
    }

    if (startDate || endDate) {
      where.recordedAt = {};
      if (startDate) where.recordedAt.gte = startDate;
      if (endDate) where.recordedAt.lte = endDate;
    }

    const measurements = await this.prisma.vitalMeasurement.findMany({
      where,
      orderBy: {
        recordedAt: 'desc',
      },
    });

    return measurements.map((m) => this.mapToResponse(m));
  }

  /**
   * Find one vital by ID
   */
  async findOne(context: ActorContext, vitalId: bigint) {
    const measurement = await this.prisma.vitalMeasurement.findFirst({
      where: {
        vitalMeasurementId: vitalId,
        elderUserId: context.elderUserId,
      },
    });

    if (!measurement) {
      throw new NotFoundException('Vital measurement not found');
    }

    return this.mapToResponse(measurement);
  }

  /**
   * Update vital measurement
   */
  async update(context: ActorContext, vitalId: bigint, updateDto: UpdateVitalDto) {
    const measurement = await this.prisma.vitalMeasurement.findFirst({
      where: {
        vitalMeasurementId: vitalId,
        elderUserId: context.elderUserId,
      },
    });

    if (!measurement) {
      throw new NotFoundException('Vital measurement not found');
    }

    const updateData: any = {};

    if (updateDto.type !== undefined) {
      updateData.kindCode = this.typeToKindCode(updateDto.type);
      updateData.unitCode = this.getUnitCode(updateDto.type);
    }

    if (updateDto.value !== undefined) {
      const type = updateDto.type || this.kindCodeToType(measurement.kindCode);
      const parsed = this.parseValue(type, updateDto.value);
      updateData.value1 = parsed.value1;
      updateData.value2 = parsed.value2;
      updateData.valueText = parsed.valueText;
    }

    if (updateDto.timestamp !== undefined) {
      updateData.recordedAt = new Date(updateDto.timestamp);
    }

    if (updateDto.notes !== undefined) {
      updateData.notes = updateDto.notes;
    }

    const updated = await this.prisma.vitalMeasurement.update({
      where: { vitalMeasurementId: vitalId },
      data: updateData,
    });

    return this.mapToResponse(updated);
  }

  /**
   * Delete vital measurement
   */
  async remove(context: ActorContext, vitalId: bigint) {
    const measurement = await this.prisma.vitalMeasurement.findFirst({
      where: {
        vitalMeasurementId: vitalId,
        elderUserId: context.elderUserId,
      },
    });

    if (!measurement) {
      throw new NotFoundException('Vital measurement not found');
    }

    await this.prisma.vitalMeasurement.delete({
      where: { vitalMeasurementId: vitalId },
    });

    return { message: 'Vital measurement deleted successfully' };
  }

  /**
   * Get latest vitals per kind (use database view)
   */
  async getLatest(context: ActorContext) {
    // Use raw query to access the view
    const results = await this.prisma.$queryRaw`
      SELECT * FROM "v_vitals_latest_per_kind"
      WHERE "elderUserId" = ${context.elderUserId}
    `;

    return (results as any[]).map((r) => this.mapToResponse(r));
  }

  /**
   * Get trends for vital measurements
   */
  async getTrends(context: ActorContext, kindCode?: string, days: number = 7) {
    const cutoffDate = new Date();
    cutoffDate.setDate(cutoffDate.getDate() - days);

    const where: any = {
      elderUserId: context.elderUserId,
      recordedAt: {
        gte: cutoffDate,
      },
    };

    if (kindCode) {
      where.kindCode = kindCode;
    }

    const measurements = await this.prisma.vitalMeasurement.findMany({
      where,
      orderBy: {
        recordedAt: 'asc',
      },
    });

    if (measurements.length === 0) {
      return kindCode
        ? [
            {
              kindCode,
              average: 0,
              count: 0,
              hasAbnormal: false,
              measurements: [],
            },
          ]
        : [];
    }

    // Group by kindCode
    const grouped = new Map<string, any[]>();
    for (const m of measurements) {
      if (!grouped.has(m.kindCode)) {
        grouped.set(m.kindCode, []);
      }
      grouped.get(m.kindCode)!.push(m);
    }

    const results: any[] = [];

    for (const [code, vitals] of grouped.entries()) {
      // Calculate average (handle different value types)
      let sum = 0;
      let count = 0;
      const hasAbnormal = vitals.some((v) => this.isAbnormal(v));

      for (const vital of vitals) {
        if (vital.value1 !== null) {
          sum += parseFloat(vital.value1.toString());
          count++;
        } else if (vital.valueText) {
          // For blood pressure, parse "120/80" format
          const parts = vital.valueText.split('/');
          if (parts.length === 2) {
            const systolic = parseFloat(parts[0]);
            if (!isNaN(systolic)) {
              sum += systolic;
              count++;
            }
          }
        }
      }

      const average = count > 0 ? sum / count : 0;

      results.push({
        kindCode: code,
        average: Math.round(average * 100) / 100,
        count: vitals.length,
        hasAbnormal,
        measurements: vitals.map((v) => this.mapToResponse(v)),
      });
    }

    return results;
  }

  /**
   * Get abnormal readings
   */
  async getAbnormal(context: ActorContext) {
    const measurements = await this.prisma.vitalMeasurement.findMany({
      where: {
        elderUserId: context.elderUserId,
      },
      orderBy: {
        recordedAt: 'desc',
      },
    });

    return measurements
      .filter((m) => this.isAbnormal(m))
      .map((m) => this.mapToResponse(m));
  }

  /**
   * Check if a reading is abnormal
   */
  private isAbnormal(measurement: any): boolean {
    const kindCode = measurement.kindCode.toLowerCase();
    const value1 = measurement.value1 ? parseFloat(measurement.value1.toString()) : null;
    const value2 = measurement.value2 ? parseFloat(measurement.value2.toString()) : null;

    if (kindCode === 'bp' && value1 && value2) {
      // Blood pressure: >140/90 or <80/50
      return value1 > 140 || value1 < 80 || value2 > 90 || value2 < 50;
    }

    if (kindCode === 'bs' && value1 !== null) {
      // Blood sugar: >125 or <60
      return value1 > 125 || value1 < 60;
    }

    if (kindCode === 'hr' && value1 !== null) {
      // Heart rate: <50 or >110
      return value1 < 50 || value1 > 110;
    }

    if (kindCode === 'temp' && value1 !== null) {
      // Temperature: <96.0 or >100.4
      return value1 < 96.0 || value1 > 100.4;
    }

    if (kindCode === 'o2' && value1 !== null) {
      // Oxygen saturation: <90
      return value1 < 90;
    }

    return false;
  }

  /**
   * Get unit code for vital type
   */
  private getUnitCode(type: VitalType): string {
    const mapping: Record<VitalType, string> = {
      [VitalType.BLOOD_PRESSURE]: 'mmHg',
      [VitalType.BLOOD_SUGAR]: 'mg/dL',
      [VitalType.HEART_RATE]: 'bpm',
      [VitalType.TEMPERATURE]: '°F',
      [VitalType.OXYGEN_SATURATION]: '%',
      [VitalType.WEIGHT]: 'lbs',
    };
    return mapping[type] || '';
  }

  /**
   * Map database model to API response
   */
  private mapToResponse(measurement: any) {
    return {
      id: measurement.vitalMeasurementId.toString(),
      type: this.kindCodeToType(measurement.kindCode),
      value: this.formatValue(measurement),
      timestamp: measurement.recordedAt.toISOString(),
      notes: measurement.notes || null,
      userId: measurement.elderUserId.toString(),
    };
  }
}

